### PR TITLE
Add configuration message.

### PR DESCRIPTION
**Pull request by steveluc**
Added configuration message.  Added logic to expand tabs to spaces using host-configured tab size.

### CODE REVIEW

File: src/server/editorServices.ts [hunk 1]
```diff
@@ -609,6 +637,7 @@ module ts.server {
                     }
                 }
                 if (content !== undefined) {
+                    content = this.expandTabs(content);
```
Comment (CyrusNajmabadi) [inline src/server/editorServices.ts]: Is this changing the data that the host gave us?  If so, won't our lines/characters be off in the responses we send the host?

Comment (mhegazy) [inline src/server/editorServices.ts]: I think this goes to show that loading the file from disk does not seem to be the right way. Ideally the openfile message to send the contents. the issue is perf when doing the reload, assuming that this is a common occurrence.

File: src/server/protocol.d.ts [hunk 2]
```diff
@@ -300,6 +300,36 @@ declare module ts.server.protocol {
     }
 
     /**
+      * Information found in a configure request.
+      */
+    export interface ConfigureRequestArguments {
+        /** Number of spaces for each tab */
+        tabSize: number;
```
Comment (mhegazy) [inline src/server/protocol.d.ts]: what about NewLineCharacter, and ConvertTabsToSpaces, why are not they included here?

**Comment by mhegazy** [comment]
:+1:

**Comment by steveluc** [comment]
On ConvertTabsToSpaces, it is more complex and error prone on the server side to maintain the tabs because it breaks the invariant of one character to one column.  So far I don't see an upside because the server's representation is never saved to disk or viewed by the editor user.  Therefore on the server we always convert tabs to spaces.  If the tab size changes in the editor, it needs to notify the server and also to send a reload message to the server.  

On NewLine, I think we should add that to the global message and also have a per file message (with NewLine and TabSize).

**Comment by CyrusNajmabadi** [comment]
> On ConvertTabsToSpaces, it is more complex and error prone on the server side to maintain the tabs because it breaks the invariant of one character to one column.

Why is this invariant important?  In VS we found that if you use different representations for the buffer, you're going to have an unending tail of issues you run into due to you never quite getting the internal representation right.  A long time ago we switched over to just using the host's view of the file, and eliminated all those issues.  

It seems like we're going down that path again.  It would probably be *much* simpler (both in code and concept) to just get the buffer from the tool, operate only on that representation (including producing all answers only with that representation), than to try to continually get all the host editor semantics mapped perfectly into our side.  

Doing the latter also concerns me because we don't just provide answers to simple read-only queries.  We also do things like tell an editor what sort of edits to make.  When the representation is inaccurate that can lead to data corruption if we provide errant edits.

I've done some perf tests and pipes seem exceedingly fast and more than good enough for this.  On my windows box i got us able to send a 10MB of text down the pipe in 0.15 seconds.

**Comment by steveluc** [comment]
@CyrusNajmabadi I am sympathetic to the argument you are making.  Theoretically, it is the right argument.  In practice, making the change now to put tabs into the server would touch many more code locations than just expanding the tabs and the algorithm would end up being the same: count each tab as TabSize columns, communicate back to the editor in terms of column number.  

Let's plan to revisit this post-release or in a branch so that we have time to stabilize it.

**Comment by steveluc** [comment]
@CyrusNajmabadi on the pipe speed, that is not the concern.  The concern is JSON encoding speed in python and also fidelity of the roundtrip through JSON encoding.  Would love to see some experiments on this.

**Comment by CyrusNajmabadi** [comment]
+



+        expandTabs(text: string) {


+            var spaces = generateSpaces(this.hostConfiguration.formatCodeOptions.TabSize);


+            return text.replace(/\t/g, spaces);


+        }



Note: this approach to expanding out tabs isn’t necessarily correct either.  VS, for example, doesn’t treat a tab as a
certain number of spaces.  What it does is use tab to align to the next column that is a multiple of whatever your tab size is.  i.e. if you have a tabsize of ‘4’, and you have “space space tab” then that is *not* equivalent to a string with six spaces in it.  Instead, it’s equivalent of having four spaces.  You can see this here in VS:

[cid:image001.png@01D06127.D7481A70]

Here both ‘class’ and then open brace are on column 4.

If we can presume that all our hosts behave in the same manner, then we’ll have to do things differently in order to properly convert the text we’ve read into a space-to-column oriented model that matches them.  We’ll first have to convert the text to lines first, then go through each line character by character until we hit tabs.  Then we will compute how many are necessary such that a tab in this location would align us with the right column that is a multiple of the tab size.

             -- Cyrus

**Pull request merged**